### ISSUE TITLE
Figure out how to support sdist command

### ISSUE THREAD

**Issue by chshersh** [kowainik/autopack#2]

**Comment by hasufell** [comment]
This is a bigger problem than it sounds, imo, because `cabal install` uses `sdist`.

@phadej any ideas?

**Comment by hasufell** [comment]
I have a crazy idea... what if autopack would add the exposed modules during setup to the cabal file :>

It could be guarded with an environment variable, so it doesn't clutter the git history for non-sdist workflows. Writing to a file is always problematic, requires committing it, can fail etc, so I think it would be reasonable to have this optional.

**Comment by vrom911** [comment]
Interesting idea, though I am not sure that it would work as I am not sure on which stage `sdist` reads cabal file.

I guess we need to investigate more on the `UserHooks` data type. There is the `preCopy` field in there which may be something to do with `sdist`. So probably we can just override behaviour there as well 🤔 What do you think @chshersh ?

**Comment by chshersh** [comment]
@vrom911, I agree with your solution. Ideally, I'd love to see something like `sdistHook` in `UserHooks`, but if it's already possible to reuse one of the existing hooks to achieve the desired behaviour, it's worth investigating.

**Comment by hasufell** [comment]
Well, I made it work with: https://www.haskell.org/cabal/users-guide/developing-packages.html#pkg-field-extra-source-files

So if you have `hs-source-dirs:  lib`, then you just add `extra-source-files: lib/**/*.hs`. But that still has the problem that it circumvents the current .cabal file standard. I think that needs to be discussed with cabal maintainers.

As autopack doesn't really conflict with cabal-fmt, my current workflow is basically: use autopack with the above hack. Once I need to upload to hackage, I make sure my .cabal file has the modules listed explicitly (via cabal-fmt or so), while still keeping autopack in place.
